mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -544,6 +545,7 @@ impl TendermintCoin {
             history_sync_state: Mutex::new(history_sync_state),
             client: TendermintRpcClient(AsyncMutex::new(client_impl)),
             chain_registry_name: protocol_info.chain_registry_name,
+            ctx: ctx.clone(),
```

Comment:
It introduces circular reference: `MmArc -> CoinsContext -> TendermintCoin -> MmArc`. What do you think about using channels and `balance_event_tx: Sender<BalanceEvent>` in the coin instead? It should give us more flexibility on how these events are handled. Possibly, it can be also used in the future to automatically update orders if available balance is decreased: 
https://github.com/KomodoPlatform/komodo-defi-framework/blob/3c078b6a2aed2e4f9e534907e77deca466df091f/mm2src/mm2_main/src/lp_ordermatch.rs#L1096

Reply:
Do you think it's a problem? Channels will add much more overhead than incrementing the ref counter in smart pointer.

Reply:
It is not simply `incrementing the ref counter in smart pointer`. As per https://doc.rust-lang.org/std/sync/struct.Arc.html#breaking-cycles-with-weak:
> A cycle between Arc pointers will never be deallocated.

So, on platforms where MM2 doesn't run as separate process (WASM and mobile), `MmArc` won't be dropped when `stop` is invoked, if Tendermint coin is activated. 

I pushed a demo of the problem: https://github.com/KomodoPlatform/komodo-defi-framework/commit/0f9a46cb1225100d2fd39d7123b9e482a681fdee. In this commit, `test_mm2_stops_immediately` WASM test fails. If you comment out IRIS-TEST activation part, the test will pass.

AFAIK, GUIs may need to restart MM2 without reloading the page/stopping the entire application process. Having this cycle will likely cause them to hang, waiting indefinitely for MmArc to be dropped.

Also, having partially running previous instances caused serious issues in the past. E.g., 2 instances of the same swap could be started, making a node sending their HTLC transaction twice. It was a reason to have this PR: https://github.com/KomodoPlatform/komodo-defi-framework/pull/628.

Reply:
I see now, thank you for the great explanation!

Reply:
Used weak pointer instead; the test passes as expected now. https://github.com/KomodoPlatform/komodo-defi-framework/actions/runs/6429238388/job/17457921618#step:7:642